### PR TITLE
Add support for compiling Objective-C code as extra sources.

### PR DESCRIPTION
**Pull request by jacob-carlborg**
Add the `EXTRA_OBJC_SOURCES` key and a list of Objective-C files to compile in a comment in the test file.

This is the first step for implementing [DIP 43 - D/Objective-C](http://wiki.dlang.org/DIP43).

### CODE REVIEW

**Comment by yebblies** [comment]
Why do the objc tests need their own subfolders?

**Comment by yebblies** [comment]
`EXTRA_OBJC_SOURCES` should probably do the same thing as `DISABLED` on non-apple platforms.

**Comment by jacob-carlborg** [comment]
> Why do the objc tests need their own subfolders?

No reason, other than to organize the files. There will be a bunch of files in the end. I don't have the exact numbers but around 30 files.

**Comment by jacob-carlborg** [comment]
> `EXTRA_OBJC_SOURCES` should probably do the same thing as `DISABLED` on non-apple platforms.`

I don't plan to run the tests for these files at all on other platforms. Do yo want it as an extra precaution?

**Comment by yebblies** [comment]
> No reason, other than to organize the files. There will be a bunch of files in the end. I don't have the exact numbers but around 30 files.

I'm not sure 30 files warrants the complexity of three folders.  An `objc_` prefix should be plenty to identify them.

> I don't plan to run the tests for these files at all on other platforms. Do yo want it as an extra precaution?
>
> The way I plan this to work is to have a D_OBJC variable in the Makefile. D_OBJC will by default be enabled for OS X. The Objective-C related tests will only run if D_OBJC is enabled.

Oh right the makefile.  I forget about it because I don't use it on win32.  I'm hoping to get rid of it and move the logic into `d_do_test` so that cygwin is not required.  Hopefully this makes my request make more sense.

Also, is the autotester ready to compile the tests on osx?  If so please add a basic objc example test.

**Comment by jacob-carlborg** [comment]
> Oh right the makefile. I forget about it because I don't use it on win32. I'm hoping to get rid of it and move the logic into d_do_test so that cygwin is not required. Hopefully this makes my request make more sense.

If that's the case, I would assume the D version would process the command line arguments to allow a similar syntax as make. i.e. `d_do_test D_OBJC=1`.

I do have another branch that allows one to specify `PLATFORM: osx`, for example, to make a test only run on a given platform. I used that approach first but then figured using ` D_OBJC` was easier. Note, `EXTRA_OBJC_SOURCES` is only required on very few tests.

> Also, is the autotester ready to compile the tests on osx? If so please add a basic objc example test

Yes. Do you mean like adding an empty tests that compiles some Objective-C code using `EXTRA_OBJC_SOURCES`?

**Comment by jacob-carlborg** [comment]
I think having separate folders might make it easier to ignore them on other platforms using `D_OBJC`.

**Comment by jacob-carlborg** [comment]
@WalterBright @9rnsr any comments?

**Comment by yebblies** [comment]
Is there any reason _not_ to run the tests on OSX?  You could also use an empty `EXTRA_OBJC_SOURCES` to mark them as objc tests.

> Yes. Do you mean like adding an empty tests that compiles some Objective-C code using EXTRA_OBJC_SOURCES?

Yeah.

**Comment by jacob-carlborg** [comment]
> Is there any reason not to run the tests on OSX?

As per request I'm only going to implement dobjc for 64bit. Or rather the modern runtime, it should in theory work on iOS.

> You could also use an empty EXTRA_OBJC_SOURCES to mark them as objc tests.

I was hoping to not have to tag every single test file.



>> Yes. Do you mean like adding an empty tests that compiles some Objective-C code using >>EXTRA_OBJC_SOURCES?

> Yeah.

I'll do that.

**Comment by jacob-carlborg** [comment]
@yebblies I've updated the pull request according to you're comments:

* The objc tests don't use their own subfolders
* If the variable `EXTRA_OBJC_SOURCES` is present in a test, the test will be ignored unless the `D_OBJC` environment variable is `1`. `D_OBJC` is by default `1` on OS X 64bit

**Comment by yebblies** [comment]
Ok it looks good, but the autotester is failing.

What's the deal with the postscript?  It doesn't seem to be referenced.

**Comment by jacob-carlborg** [comment]
> Ok it looks good, but the autotester is failing.

Hmm, I'll see what I can figure out.

> What's the deal with the postscript? It doesn't seem to be referenced.

I thought that was handled automatically. All I want is to make sure the Objective-C object files is properly cleaned up. Is that handled automatically?

**Comment by yebblies** [comment]
I have no idea.  What happens when you run it locally?  I expect you need to add a `POST_SCRIPT:` comment if you want the postscript to actually run, but it would probably be better to mimic what the c++ support does instead of requiring that in every file.

**Comment by yebblies** [comment]
If the postscript is supposed to be deleting the D object, then it's almost certainly redundant.  If it's meant to be doing the objc one, then it looks like it's the wrong filename: `clang -m64 -c runnable/extra-files/objc_self_test.m -o generated/runnable/objc_self_test.m.o`

**Comment by jacob-carlborg** [comment]
I see now that the Objective-C object file is placed in `test_results`, which is removed when running `make clean`. I'll remove the post script.

**Comment by jacob-carlborg** [comment]
I removed the post script and I think I know what didn't work on the auto tester.

**Comment by jacob-carlborg** [comment]
Everything is green now.

File: test/d_do_test.d [hunk 1]
```diff
@@ -375,9 +430,9 @@ int main(string[] args)
 
     switch (input_dir)
     {
-        case "compilable":       testArgs.mode = TestMode.COMPILE;      break;
-        case "fail_compilation": testArgs.mode = TestMode.FAIL_COMPILE; break;
-        case "runnable":         testArgs.mode = TestMode.RUN;          break;
+        case "compilable":              testArgs.mode = TestMode.COMPILE;      break;
```
Comment (yebblies) [inline test/d_do_test.d]: Leftover change?

**Comment by yebblies** [comment]
Auto-merge toggled on

File: test/d_do_test.d (hunk 1 above)
Comment (jacob-carlborg) [inline test/d_do_test.d]: Yes, it looks like that.

**Comment by jacob-carlborg** [comment]
> Auto-merge toggled on

Thanks.

**Pull request merged**